src/kOS/Execution/Stack.cs

```diff
@@ -48,9 +52,19 @@ private void ThrowIfInvalid(object item)
         /// <returns></returns>
         private object ProcessItem(object item)
         {
-            if (item is float)
+            if (item is float) {
                 // promote floats to doubles
-                return Convert.ToDouble(item);
+                item = Convert.ToDouble (item);
+            }
+            if (item is double)
+            {
+                if (!Double.IsNaN((double)item) && Int32.MinValue < (double)item && (double)item < Int32.MaxValue)
+                {
+                    int intPart = Convert.ToInt32(item);
+                    if ((double)item == (double)intPart)
+                        item = intPart;
+                }
+            }
```

Comment:
It's not a matter of expense but of when.  The idea is to fix all the spots where an int is presumed by putting the fix in one place instead of in 20 places.  But I'm still unsure if that's the right thing to do, because then what about all the places that expect a double and now will be getting an int?  I think it's inevitable that you'll end up having to perform that cast on a per-instance basis in the C# code, because there are cases where an int is right and cases where a double is right, so maybe trying to solve it in one universal place isn't really practical.